File: problem/Longest_Palindromic_Substring/longestPalindrome.js
/**
 * Expand around centers - O(n^2) time, O(1) space
 */
var longestPalindrome = function(s) {
    if (!s || s.length === 0) return "";
    
    let start = 0, maxLen = 1;
    for (let i = 0; i < s.length; i++) {
        const len1 = expandAroundCenter(s, i, i); 
        const len2 = expandAroundCenter(s, i, i + 1); 
        
        const currentMax = Math.max(len1, len2);
        if (currentMax > maxLen) {
            maxLen = currentMax;
            start = i - Math.floor((currentMax - 1) / 2);
        }
    }
    
    return s.substring(start, start + maxLen);
};

function expandAroundCenter(s, left, right) {
    while (left >= 0 && right < s.length && s[left] === s[right]) {
        left--;
        right++;
    }
    return right - left - 1;
}

/**
 * Dynamic programming approach - O(n^2) time, O(n^2) space
 */
var longestPalindrome_dp = function(s) {
    const n = s.length;
    if (n === 0) return "";
    
    const dp = Array(n).fill().map(() => Array(n).fill(false));
    let start = 0, maxLen = 1;

    for (let i = 0; i < n; i++) {
        dp[i][i] = true;
    }

    for (let i = 0; i < n - 1; i++) {
        if (s[i] === s[i + 1]) {
            dp[i][i + 1] = true;
            start = i;
            maxLen = 2;
        }
    }

    for (let len = 3; len <= n; len++) {
        for (let i = 0; i <= n - len; i++) {
            const j = i + len - 1;
            if (s[i] === s[j] && dp[i + 1][j - 1]) {
                dp[i][j] = true;
                start = i;
                maxLen = len;
            }
        }
    }
    
    return s.substring(start, start + maxLen);
};

/**
 * Optimized expand around centers
 */
var longestPalindrome_optimized = function(s) {
    if (!s || s.length === 0) return "";
    
    let start = 0, maxLen = 1;
    for (let i = 0; i < s.length; i++) {
        if (s.length - i <= maxLen / 2) break;
        
        let left = i, right = i;
        while (right < s.length - 1 && s[right] === s[right + 1]) {
            right++;
        }
        
        i = right;

        while (left > 0 && right < s.length - 1 && s[left - 1] === s[right + 1]) {
            left--;
            right++;
        }
        
        const currentLen = right - left + 1;
        if (currentLen > maxLen) {
            maxLen = currentLen;
            start = left;
        }
    }
    
    return s.substring(start, start + maxLen);
};

/**
 * Manacher's algorithm - O(n) time, O(n) space
 */
var longestPalindrome_manacher = function(s) {
    if (!s || s.length === 0) return "";

    let processed = "^#" + s.split('').join('#') + "#$";
    const n = processed.length;
    const lengths = new Array(n).fill(0);

    let center = 0, right = 0;
    for (let i = 1; i < n - 1; i++) {
        const mirror = 2 * center - i;
        
        if (i < right) {
            lengths[i] = Math.min(right - i, lengths[mirror]);
        }
        
        while (processed[i + lengths[i] + 1] === processed[i - lengths[i] - 1]) {
            lengths[i]++;
        }

        if (i + lengths[i] > right) {
            center = i;
            right = i + lengths[i];
        }
    }

    let maxLen = 0, centerIndex = 0;
    for (let i = 1; i < n - 1; i++) {
        if (lengths[i] > maxLen) {
            maxLen = lengths[i];
            centerIndex = i;
        }
    }
    
    const start = Math.floor((centerIndex - maxLen) / 2);
    return s.substring(start, start + maxLen);
};

File: problem/Longest_Palindromic_Substring/longestPalindrome.py
class Solution(object):
    def longestPalindrome(self, s):
        """
        Expand around centers - O(n^2) time, O(1) space
        """
        if not s:
            return ""
        
        start = 0
        max_len = 1
        
        for i in range(len(s)):
            len1 = self.expandAroundCenter(s, i, i)
            len2 = self.expandAroundCenter(s, i, i + 1)
            
            current_max = max(len1, len2)
            if current_max > max_len:
                max_len = current_max
                start = i - (current_max - 1) // 2
        
        return s[start:start + max_len]
    
    def expandAroundCenter(self, s, left, right):
        """Helper function to expand around center"""
        while left >= 0 and right < len(s) and s[left] == s[right]:
            left -= 1
            right += 1
        return right - left - 1
    
    def longestPalindrome_dp(self, s):
        """
        Dynamic programming approach - O(n^2) time, O(n^2) space
        """
        n = len(s)
        if n == 0:
            return ""

        dp = [[False] * n for _ in range(n)]
        start = 0
        max_len = 1

        for i in range(n):
            dp[i][i] = True

        for i in range(n - 1):
            if s[i] == s[i + 1]:
                dp[i][i + 1] = True
                start = i
                max_len = 2

        for length in range(3, n + 1):
            for i in range(n - length + 1):
                j = i + length - 1
                if s[i] == s[j] and dp[i + 1][j - 1]:
                    dp[i][j] = True
                    start = i
                    max_len = length
        
        return s[start:start + max_len]
    
    def longestPalindrome_manacher(self, s):
        """
        Manacher's algorithm - O(n) time, O(n) space
        """
        processed = "^#" + "#".join(s) + "#$"
        n = len(processed)

        lengths = [0] * n
        center = right = 0
        
        for i in range(1, n - 1):
            mirror = 2 * center - i

            if i < right:
                lengths[i] = min(right - i, lengths[mirror])

            while processed[i + lengths[i] + 1] == processed[i - lengths[i] - 1]:
                lengths[i] += 1

            if i + lengths[i] > right:
                center, right = i, i + lengths[i]

        max_len = 0
        center_index = 0
        for i in range(1, n - 1):
            if lengths[i] > max_len:
                max_len = lengths[i]
                center_index = i

        start = (center_index - max_len) // 2
        return s[start:start + max_len]
